src/core/api.rs

```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
         )
     }
 
-    /// Generates a random symmetric key to be used with a DEM scheme and
-    /// generates its encapsulation for the given access policy.
-    // TODO document error cases.
-    pub fn encaps(
+    /// Generate a user secret key with the given rights.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn keygen(
         &self,
+        msk: &mut MasterSecretKey,
         policy: &Policy,
-        pk: &MasterPublicKey,
-        access_policy: AccessPolicy,
-    ) -> Result<(SymmetricKey<SEED_LENGTH>, Encapsulation), Error> {
+        ap: &str,
+    ) -> Result<UserSecretKey, Error> {
+        let ap = &AccessPolicy::parse(ap)?;
+        usk_keygen(
+            &mut *self.rng.lock().expect("Mutex lock failed!"),
+            msk,
+            policy.generate_semantic_space_coordinates(ap.clone())?,
+        )
+    }
+    /// Generates an encapsulation for the given access
+    /// policy.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encaps(
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+    ) -> Result<(Secret<SEED_LENGTH>, Encapsulation), Error> {
+        let ap = &AccessPolicy::parse(ap)?;
         encaps(
             &mut *self.rng.lock().expect("Mutex lock failed!"),
-            pk,
-            &policy.generate_point_coordinates(access_policy)?,
+            mpk,
+            &policy.generate_point_coordinates(ap.clone())?,
         )
     }
-
-    /// Attempts opening the given `Covercrypt` encapsulation. Returns the
-    /// encapsulated key upon success and `None` otherwise.
-    // TODO document error cases.
-    pub fn decaps(
+    /// Attempts opening the given encapsulation using the given
+    /// user secret key.
+    ///
+    /// Returns the encapsulated symmetric key if the user key holds
+    /// the correct rights.
+    fn decaps(
         &self,
         usk: &UserSecretKey,
-        encapsulation: &Encapsulation,
-    ) -> Result<Option<SymmetricKey<SEED_LENGTH>>, Error> {
-        decaps(usk, encapsulation)
+        encapsulation: Encapsulation,
+    ) -> Option<Secret<SEED_LENGTH>> {
+        match decaps(usk, &encapsulation) {
+            Err(_) => None,
+            Ok(sk) => Some(sk?),
+        }
```

Comment:
No you should keep a `Result<Option<...>, Error>`: the error semantic is not the same as the option one. You get an error in case something went wrong; you get `None` if the USK rights did not match the encapsulation's access policy.